### PR TITLE
Test and fix put broadcast

### PR DESCRIPTION
**Pull request by netsirius**
### Fix and test the broadcasting after put operation

The put operation flow has been fixed and tested. Now the broadcast to the subscribing nodes is done correctly.

The flow would be the following:
broadcast_flow

### CODE REVIEW

File: crates/locutus-node/src/node/test_utils.rs [hunk 1]
```diff
@@ -239,6 +240,27 @@ impl SimNetwork {
         });
     }
 
+    pub fn get_locations_by_node(&self) -> HashMap<String, PeerKeyLocation> {
+        let mut locations_by_node: HashMap<String, PeerKeyLocation> = HashMap::new();
+        let nodes = self.nodes.as_slice().clone();
+        let gateways = self.gateways.as_slice().clone();
+
+        // Get node and gateways location by label
+        for (node, label) in nodes {
```
Comment (iduartgomez) [inline crates/locutus-node/src/node/test_utils.rs:249]: you can do this `for ... &self.nodes` and will deref to an slice directly

File: crates/locutus-node/src/node/test_utils.rs [hunk 2]
```diff
@@ -57,8 +58,8 @@ pub(crate) type EventId = usize;
 
 #[derive(Clone)]
 pub(crate) struct NodeSpecification {
-    /// Pair of contract and the initial value
-    pub owned_contracts: Vec<(Contract, ContractValue)>,
+    /// Pair of contract, the initial value and contract subscribers
+    pub owned_contracts: Vec<(Contract, ContractValue, Vec<PeerKeyLocation>)>,
```
Comment (iduartgomez) [inline crates/locutus-node/src/node/test_utils.rs:62]: I think it's better for clarity if we split this, leave owned_contracts as it is, and then add a new field which is a map of `ContractKey <- Vec<PeerKeyLocation>`.  You can relate the subscribers to the contracts b y the contract key later.

File: crates/locutus-node/src/node/test_utils.rs [hunk 3]
```diff
@@ -239,6 +240,27 @@ impl SimNetwork {
...
+        let gateways = self.gateways.as_slice().clone();
+
+        // Get node and gateways location by label
+        for (node, label) in nodes {
+            locations_by_node.insert(
+                label.to_string(),
+                node.op_storage.ring.own_location().clone(),
+            );
+        }
+        for (node, config) in gateways {
```
Comment (iduartgomez) [inline crates/locutus-node/src/node/test_utils.rs:255]: likewise for gateways

File: crates/locutus-node/src/node/event_listener.rs [hunk 4]
```diff
@@ -167,6 +188,13 @@ fn create_log(logs: &[MessageLog], log: EventLog) -> (MessageLog, ListenerLogId)
 }
 
 fn fuse_events_msg<'a>(mut put_ops: impl Iterator<Item = &'a EventKind>) -> Option<EventKind> {
```
Comment (iduartgomez) [inline crates/locutus-node/src/node/event_listener.rs:190]: let's rename the function name to `fuse_successful_put_op` for clarity

File: crates/locutus-node/src/node/event_listener.rs [hunk 5]
```diff
@@ -167,6 +188,13 @@ fn create_log(logs: &[MessageLog], log: EventLog) -> (MessageLog, ListenerLogId)
...
+    let mut put_ops = put_ops.filter(|op| {
```
Comment (iduartgomez) [inline crates/locutus-node/src/node/event_listener.rs:191]: do the filtering in the iterator being passed to the function (which is already filtering itself)

File: crates/locutus-node/src/node/event_listener.rs [hunk 6]
```diff
@@ -234,6 +262,17 @@ mod test_utils {
             })
         }
 
+        pub fn get_broadcast_count(
+            &self,
+            expected_key: &ContractKey,
+            expected_value: &ContractValue,
+        ) -> usize {
+            let mut logs = self.logs.read();
+            logs.iter().filter(|log| {
```
Comment (iduartgomez) [inline crates/locutus-node/src/node/event_listener.rs:271]: if we are doing this for testing pourpouses, wouldn't it be more correct to check that the broadcast has been received by the target nodes instead?

File: crates/locutus-node/src/operations/put.rs [hunk 7]
```diff
@@ -426,6 +436,65 @@ where
...
+            if let PutMsg::SuccessfulUpdate { .. } = internal_cb {
+                log::debug!(
+                    "Empty broadcast list while updating value for contract {}",
+                    key
+                );
+                // means the whole tx finished so can return early
+                return_msg = Some(internal_cb.into());
+                new_state = None;
+            } else {
+                log::debug!("Notify nodes to to broadcast");
```
Comment (iduartgomez) [inline crates/locutus-node/src/operations/put.rs:491]: typo: double "to"
Also more correct would be "callback to start broadcasting to other nodes"

**Comment by iduartgomez** [comment]
The PR is not compiling due to an import/dependency.

File: crates/locutus-node/src/node/event_listener.rs (hunk 6 above)
Comment (netsirius) [reply crates/locutus-node/src/node/event_listener.rs:271]: Yes, I had thought to do that, the problem is that right now the target node of the broadcast is random. Another option that I have thought of is to verify that the target node is one of the two nodes to which should be broadcast.

**Comment by netsirius** [comment]
@iduartgomez I have resolved most of the comments. Could you check it out?

File: crates/locutus-node/src/node/event_listener.rs (hunk 6 above)
Comment (iduartgomez) [reply crates/locutus-node/src/node/event_listener.rs:267]: Yeah is enough to verify that any node (whichever it is) has received the broadcast (for now at least, eventually we would have to verify that multiple broadcasts have happened for different nodes when initiating one put). If you can change that we can merge this.

File: crates/locutus-node/src/node/test_utils.rs [hunk 8]
```diff
@@ -57,10 +58,11 @@ pub(crate) type EventId = usize;
 
 #[derive(Clone)]
 pub(crate) struct NodeSpecification {
-    /// Pair of contract and the initial value
+    /// Pair of contract, the initial value and contract subscribers
```
Comment (iduartgomez) [inline crates/locutus-node/src/node/test_utils.rs:61]: this comment is outdated now, does not contain the subscribers

**Comment by iduartgomez** [comment]
The clippy and fmt checks should pass at least (there are a couple unit tests which are not fixed yet but we can ignore those for now). Rememebr to run `cargo fmt` and check the clippy error.

**Review by iduartgomez — APPROVED** [review]

**Pull request merged**